hashes/src/sha256.rs

```diff
@@ -134,33 +146,55 @@ impl Hash {
+///
+/// The `Midstate` type is obscure and specialized and should not be used unless you are sure of
+/// what you are doing.
+///
+/// It represents "partially hashed data" but does not itself have properties of cryptographic
+/// hashes. For example, when (ab)used as hashes, midstates are vulnerable to trivial
+/// length-extension attacks. They are typically used to optimize the computation of full hashes.
+/// For example, when implementing BIP-340 tagged hashes, which always begin by hashing the same
+/// fixed 64-byte prefix, it makes sense to hash the prefix once, store the midstate as a constant,
+/// and hash any future data starting from the constant rather than from a fresh hash engine.
+///
+/// For BIP-340 support we provide the [`sha256t`] module, and the [`sha256t_hash_newtype`]
+/// macro which will create the midstate for you in const context and use it for tagged hashing.
+#[derive(Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Midstate {
+    /// Raw bytes of the midstate i.e., the already-hashed contents of the hash engine.
+    bytes: [u8; 32],
+    /// Number of bytes hashed to achieve this midstate.
+    // INVARIANT must always be a multiple of 64.
+    length: usize,
+}
 
 impl Midstate {
-    /// Length of the midstate, in bytes.
-    const LEN: usize = 32;
-
-    /// Construct a new [`Midstate`] from the inner value.
-    pub const fn from_byte_array(inner: [u8; 32]) -> Self { Midstate(inner) }
-
-    /// Copies a byte slice into the [`Midstate`] object.
-    pub fn from_slice(sl: &[u8]) -> Result<Midstate, FromSliceError> {
-        if sl.len() != Self::LEN {
-            Err(FromSliceError { expected: Self::LEN, got: sl.len() })
-        } else {
-            let mut ret = [0; 32];
-            ret.copy_from_slice(sl);
-            Ok(Midstate(ret))
+    /// Construct a new [`Midstate`] from the `state` and the `bytes_hashed` to get to that state.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `bytes_hashed` is not a multiple of 64.
+    pub const fn new(state: [u8; 32], bytes_hashed: usize) -> Self {
+        if bytes_hashed % 64 != 0 {
+            // When MSRV is 1.57+ we can use `panic!()`.
+            #[allow(unconditional_panic)]
+            #[allow(clippy::let_unit_value)]
+            #[allow(clippy::out_of_bounds_indexing)]
+            let _panic = [(); 0][1];
         }
+
+        Midstate { bytes: state, length: bytes_hashed }
     }
 
-    /// Unwraps the [`Midstate`] and returns the underlying byte array.
-    pub const fn to_byte_array(self) -> [u8; 32] { self.0 }
+    /// Deconstructs the [`Midstate`], returning the underlying byte array and number of bytes hashed.
+    pub const fn as_parts(&self) -> (&[u8; 32], usize) { (&self.bytes, self.length) }
+
+    /// Deconstructs the [`Midstate`], returning the underlying byte array and number of bytes hashed.
+    pub const fn into_parts(self) -> ([u8; 32], usize) { (self.bytes, self.length) }
```

Comment:
The type is `Copy`, so it should be `to_parts`.

Reply:
Ah correct, tracking issue created!